File: backend/routes/reviews.py
"""
TravelVista — Reviews Routes
"""
import time
from datetime import date
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import get_db, review_to_dict

reviews_bp = Blueprint('reviews', __name__, url_prefix='/api/reviews')


def _is_admin(uid):
    conn = get_db()
    row = conn.execute("SELECT role FROM users WHERE id = ?", (uid,)).fetchone()
    conn.close()
    return row and row['role'] == 'admin'


@reviews_bp.route('/recent', methods=['GET'])
def recent_reviews():
    """Public: latest reviews for testimonials."""
    conn = get_db()
    rows = conn.execute(
        "SELECT * FROM reviews ORDER BY created_at DESC LIMIT 10"
    ).fetchall()
    conn.close()
    return jsonify([review_to_dict(r) for r in rows])


@reviews_bp.route('', methods=['GET'])
@jwt_required()
def list_reviews():
    """Admin: all reviews."""
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    rows = conn.execute("SELECT * FROM reviews ORDER BY created_at DESC").fetchall()
    conn.close()
    return jsonify([review_to_dict(r) for r in rows])


@reviews_bp.route('/package/<pkg_id>', methods=['GET'])
def package_reviews(pkg_id):
    """Public: reviews for a specific package."""
    conn = get_db()
    rows = conn.execute(
        "SELECT * FROM reviews WHERE package_id = ? ORDER BY created_at DESC",
        (pkg_id,)
    ).fetchall()
    conn.close()

    reviews = [review_to_dict(r) for r in rows]
    avg = 0
    if reviews:
        avg = round(sum(r['rating'] for r in reviews) / len(reviews), 1)

    return jsonify({'reviews': reviews, 'averageRating': avg})


@reviews_bp.route('', methods=['POST'])
@jwt_required()
def add_review():
    uid = get_jwt_identity()
    data = request.get_json() or {}

    conn = get_db()
    user = conn.execute("SELECT * FROM users WHERE id = ?", (uid,)).fetchone()
    if not user:
        conn.close()
        return jsonify({'message': 'User not found'}), 404

    rev_id = 'rev-' + str(int(time.time() * 1000))
    conn.execute(
        """INSERT INTO reviews
           (id,user_id,user_name,user_avatar,package_id,package_title,rating,comment,created_at)
           VALUES (?,?,?,?,?,?,?,?,?)""",
        (
            rev_id, uid, user['name'], user['avatar'],
            data.get('packageId', ''),
            data.get('packageTitle', ''),
            data.get('rating', 5),
            data.get('comment', ''),
            date.today().isoformat(),
        )
    )

    # Update package review count and average rating
    pkg_id = data.get('packageId', '')
    if pkg_id:
        all_reviews = conn.execute(
            "SELECT rating FROM reviews WHERE package_id = ?", (pkg_id,)
        ).fetchall()
        count = len(all_reviews)
        avg = round(sum(r['rating'] for r in all_reviews) / count, 1) if count else 0
        conn.execute(
            "UPDATE packages SET review_count = ?, rating = ? WHERE id = ?",
            (count, avg, pkg_id)
        )

    conn.commit()
    row = conn.execute("SELECT * FROM reviews WHERE id = ?", (rev_id,)).fetchone()
    conn.close()
    return jsonify(review_to_dict(row)), 201


@reviews_bp.route('/<rev_id>', methods=['DELETE'])
@jwt_required()
def delete_review(rev_id):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    # Get the review's package_id before deleting to update counts
    review = conn.execute("SELECT * FROM reviews WHERE id = ?", (rev_id,)).fetchone()
    if review:
        conn.execute("DELETE FROM reviews WHERE id = ?", (rev_id,))
        pkg_id = review['package_id']
        all_reviews = conn.execute(
            "SELECT rating FROM reviews WHERE package_id = ?", (pkg_id,)
        ).fetchall()
        count = len(all_reviews)
        avg = round(sum(r['rating'] for r in all_reviews) / count, 1) if count else 0
        conn.execute(
            "UPDATE packages SET review_count = ?, rating = ? WHERE id = ?",
            (count, avg, pkg_id)
        )
        conn.commit()

    conn.close()
    return jsonify({'success': True})

File: api/reviews.php
<?php
/**
 * TravelVista — Reviews API
 * GET    ?action=recent        — Latest reviews (public)
 * GET                          — All reviews (admin)
 * GET    ?action=package&id=X  — Reviews for a package (public)
 * POST                         — Add review
 * DELETE ?id=X                 — Delete review (admin)
 */

header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, DELETE, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') { http_response_code(200); exit; }

require_once __DIR__ . '/../config/jwt_helper.php';

$method = $_SERVER['REQUEST_METHOD'];
$action = $_GET['action'] ?? '';
$id = $_GET['id'] ?? '';
if (!$action && !$id) {
    $uri = $_SERVER['REQUEST_URI'] ?? '';
    if (preg_match('#/api/reviews/recent#', $uri)) { $action = 'recent'; }
    elseif (preg_match('#/api/reviews/package/(\d+)#', $uri, $m)) { $action = 'package'; $id = $m[1]; }
    elseif (preg_match('#/api/reviews/(\d+)#', $uri, $m)) { $id = $m[1]; }
}

// ── GET: Recent reviews (public) ──
if ($method === 'GET' && $action === 'recent') {
    $db = getDB();
    $rows = $db->query("SELECT * FROM reviews ORDER BY created_at DESC LIMIT 10")->fetchAll();
    json_response(array_map('review_to_dict', $rows));
}

// ── GET: Package reviews (public) ──
if ($method === 'GET' && $action === 'package' && $id) {
    $db = getDB();
    $stmt = $db->prepare("SELECT * FROM reviews WHERE package_id = ? ORDER BY created_at DESC");
    $stmt->execute([$id]);
    $rows = $stmt->fetchAll();
    $reviews = array_map('review_to_dict', $rows);

    $avg = 0;
    if (count($reviews) > 0) {
        $sum = array_sum(array_column($reviews, 'rating'));
        $avg = round($sum / count($reviews), 1);
    }

    json_response(['reviews' => $reviews, 'averageRating' => $avg]);
}

// ── GET: All reviews (admin) ──
if ($method === 'GET' && !$action && !$id) {
    require_admin();
    $db = getDB();
    $rows = $db->query("SELECT * FROM reviews ORDER BY created_at DESC")->fetchAll();
    json_response(array_map('review_to_dict', $rows));
}

// ── POST: Add review ──
if ($method === 'POST') {
    $userId = require_auth();
    $data = get_json_body();
    $db = getDB();

    $stmt = $db->prepare("SELECT * FROM users WHERE id = ?");
    $stmt->execute([$userId]);
    $user = $stmt->fetch();
    if (!$user) json_response(['message' => 'User not found'], 404);

    $stmt = $db->prepare(
        "INSERT INTO reviews (user_id, user_name, user_avatar, package_id, package_title, rating, comment, created_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)"
    );
    $stmt->execute([
        $userId,
        $user['name'],
        $user['avatar'],
        $data['packageId'] ?? 0,
        $data['packageTitle'] ?? '',
        $data['rating'] ?? 5,
        $data['comment'] ?? '',
        date('Y-m-d'),
    ]);

    $revId = $db->lastInsertId();

    // Update package review count and average rating
    $pkgId = $data['packageId'] ?? 0;
    if ($pkgId) {
        $stmt = $db->prepare("SELECT rating FROM reviews WHERE package_id = ?");
        $stmt->execute([$pkgId]);
        $allReviews = $stmt->fetchAll();
        $count = count($allReviews);
        $avg = $count > 0 ? round(array_sum(array_column($allReviews, 'rating')) / $count, 1) : 0;
        $db->prepare("UPDATE packages SET review_count = ?, rating = ? WHERE id = ?")->execute([$count, $avg, $pkgId]);
    }

    $stmt = $db->prepare("SELECT * FROM reviews WHERE id = ?");
    $stmt->execute([$revId]);
    json_response(review_to_dict($stmt->fetch()), 201);
}

// ── DELETE: Delete review (admin) ──
if ($method === 'DELETE' && $id) {
    require_admin();
    $db = getDB();

    $stmt = $db->prepare("SELECT * FROM reviews WHERE id = ?");
    $stmt->execute([$id]);
    $review = $stmt->fetch();

    if ($review) {
        $db->prepare("DELETE FROM reviews WHERE id = ?")->execute([$id]);
        $pkgId = $review['package_id'];
        $stmt = $db->prepare("SELECT rating FROM reviews WHERE package_id = ?");
        $stmt->execute([$pkgId]);
        $allReviews = $stmt->fetchAll();
        $count = count($allReviews);
        $avg = $count > 0 ? round(array_sum(array_column($allReviews, 'rating')) / $count, 1) : 0;
        $db->prepare("UPDATE packages SET review_count = ?, rating = ? WHERE id = ?")->execute([$count, $avg, $pkgId]);
    }

    json_response(['success' => true]);
}

json_response(['message' => 'Invalid request'], 400);
?>
